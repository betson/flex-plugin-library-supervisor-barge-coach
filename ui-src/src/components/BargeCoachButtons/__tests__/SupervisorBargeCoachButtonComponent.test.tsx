import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { SupervisorBargeCoachButtons } from '../SupervisorBargeCoachButtonComponent';
import BargeCoachService from '../../../utils/serverless/BargeCoachService';

jest.mock('../../../utils/serverless/BargeCoachService');
jest.mock('react-redux', () => ({
  useSelector: () => ({
    muted: true,
    barge: true,
    enableBargeinButton: true,
    coaching: true,
    enableCoachButton: true,
    privateMode: false,
  }),
  useDispatch: () => jest.fn(),
}));
jest.mock('../../..', () => ({
  isAgentCoachingPanelEnabled: jest.fn().mockReturnValue(true),
}));
describe('SupervisorBargeCoachButtons component enable to disable', () => {
  it('renders correctly', () => {
    const mockTask = {
      conference: {
        conferenceSid: 'CFxxxxxx',
        participants: [
          {
            participantType: 'worker',
            workerSid: 'flex-selector-test-id',
            callSid: 'CAxxxx',
          },
        ],
        source: {
          channelParticipants: [
            {
              participantSid: 'PAxxxx',
              type: 'supervisor',
              mediaProperties: {
                status: 'joined',
              },
              routingProperties: {
                workerSid: 'flex-selector-test-id',
              },
            },
          ],
        },
      },
    };
    const wrapper = render(<SupervisorBargeCoachButtons key="callback-component" task={mockTask} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('call coachHandleClick successfully', async () => {
    const mockTask = {
      conference: {
        conferenceSid: 'CFxxxxxx',
        participants: [
          {
            participantType: 'worker',
            workerSid: 'flex-selector-test-id',
            callSid: 'CAxxxx',
          },
        ],
        source: {
          channelParticipants: [
            {
              participantSid: 'PAxxxx',
              type: 'supervisor',
              mediaProperties: {
                status: 'joined',
              },
              routingProperties: {
                workerSid: 'flex-selector-test-id',
              },
            },
          ],
        },
      },
    };
    const { getByTestId } = render(<SupervisorBargeCoachButtons task={mockTask} />);
    const coachBtn = getByTestId('coachBtn');
    expect(coachBtn).toBeEnabled();
    await userEvent.click(coachBtn);
    //expect(CallbackService.callCustomerBack).toHaveBeenCalled();
  });
  it('call coachHandleClick error due to mismatching workerSID', async () => {
    const mockErrorTask = {
      conference: {
        conferenceSid: 'CFxxxxxx',
        participants: [
          {
            participantType: 'worker',
            workerSid: 'WKxxx',
            callSid: 'CAxxxx',
          },
        ],
        source: {
          channelParticipants: [
            {
              participantSid: 'PAxxxx',
              type: 'supervisor',
              mediaProperties: {
                status: 'joined',
              },
              routingProperties: {
                workerSid: 'WKxxx',
              },
            },
          ],
        },
      },
    };
    const { getByTestId } = render(<SupervisorBargeCoachButtons task={mockErrorTask} />);
    const coachBtn = getByTestId('coachBtn');
    expect(coachBtn).toBeEnabled();
    await userEvent.click(coachBtn);
    //expect(CallbackService.callCustomerBack).toHaveBeenCalled();
  });
  it('call coachHandleClick error due to no conferenceSID', async () => {
    const consoleSpy = jest.spyOn(console, 'log');
    const mockErrorTask = {
      conference: {},
    };
    const { getByTestId } = render(<SupervisorBargeCoachButtons task={mockErrorTask} />);
    const coachBtn = getByTestId('coachBtn');
    expect(coachBtn).toBeEnabled();
    await userEvent.click(coachBtn);
    expect(consoleSpy).toHaveBeenCalledWith('conferenceSid = null, returning');
  });
  it('call bargeHandleClick successfully', async () => {
    const updateParticipantBargeCoachSpy = jest.spyOn(BargeCoachService, 'updateParticipantBargeCoach');
    const mockTask = {
      conference: {
        conferenceSid: 'CFxxxxxx',
        participants: [
          {
            participantType: 'worker',
            workerSid: 'flex-selector-test-id',
            callSid: 'CAxxxx',
          },
        ],
        source: {
          channelParticipants: [
            {
              participantSid: 'PAxxxx',
              type: 'supervisor',
              mediaProperties: {
                status: 'joined',
              },
              routingProperties: {
                workerSid: 'flex-selector-test-id',
              },
            },
          ],
        },
      },
    };
    const { getByTestId } = render(<SupervisorBargeCoachButtons task={mockTask} />);
    const bargeBtn = getByTestId('bargeBtn');
    expect(bargeBtn).toBeEnabled();
    await userEvent.click(bargeBtn);
    expect(updateParticipantBargeCoachSpy).toHaveBeenCalled();
  });

  it('call bargeHandleClick error due to mismatching workerSID', async () => {
    const mockErrorTask = {
      conference: {
        conferenceSid: 'CFxxxxxx',
        participants: [
          {
            participantType: 'worker',
            workerSid: 'WKxxxx',
            callSid: 'CAxxxx',
          },
        ],
        source: {
          channelParticipants: [
            {
              participantSid: 'PAxxxx',
              type: 'supervisor',
              mediaProperties: {
                status: 'joined',
              },
              routingProperties: {
                workerSid: 'WKxxx',
              },
            },
          ],
        },
      },
    };
    console.log('call bargeHandleClick error due to mismatching workerSID Error 1')
    const { getByTestId } = render(<SupervisorBargeCoachButtons task={mockErrorTask} />);
    console.log('call bargeHandleClick error due to mismatching workerSID Error 2')
    const bargeBtn = getByTestId('bargeBtn');
    expect(bargeBtn).toBeEnabled();
    await userEvent.click(bargeBtn);
    //expect(CallbackService.callCustomerBack).toHaveBeenCalled();
  });
  it('call bargeHandleClick error due to no conferenceSID', async () => {
    console.log('call bargeHandleClick error due to no conferenceSID Error 1')
    const consoleSpy = jest.spyOn(console, 'log');
    console.log('call bargeHandleClick error due to no conferenceSID Error 2')
    const mockErrorTask = {};
    const { getByTestId } = render(<SupervisorBargeCoachButtons task={mockErrorTask} />);
    console.log('call bargeHandleClick error due to no conferenceSID Error 3')
    const bargeBtn = getByTestId('bargeBtn');
    expect(bargeBtn).toBeEnabled();
    await userEvent.click(bargeBtn);
    expect(consoleSpy).toHaveBeenCalledWith('conferenceSid = null, returning');
  });
});
