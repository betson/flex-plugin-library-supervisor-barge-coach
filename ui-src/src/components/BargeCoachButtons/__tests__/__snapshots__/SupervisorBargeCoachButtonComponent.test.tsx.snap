// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SupervisorBargeCoachButtons component enable to disable renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-x2lkna"
        data-paste-element="FLEX"
      >
        <div
          class="css-ceamo7"
          data-paste-element="BARGE_COACH_BUTTON_BOX"
        >
          <button
            data-testid="bargeBtn"
            icon="MuteLargeBold"
            style="width: 44px; height: 44px;"
            title="Unmute"
            variant="secondary"
          />
          <button
            disabled=""
            icon="IncomingCallBold"
            style="width: 44px; height: 44px;"
            title="Barge-Out"
            variant="primary"
          />
          <button
            data-testid="coachBtn"
            icon="DefaultAvatarBold"
            style="width: 44px; height: 44px;"
            title="Disable Coach Mode"
            variant="primary"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-x2lkna"
      data-paste-element="FLEX"
    >
      <div
        class="css-ceamo7"
        data-paste-element="BARGE_COACH_BUTTON_BOX"
      >
        <button
          data-testid="bargeBtn"
          icon="MuteLargeBold"
          style="width: 44px; height: 44px;"
          title="Unmute"
          variant="secondary"
        />
        <button
          disabled=""
          icon="IncomingCallBold"
          style="width: 44px; height: 44px;"
          title="Barge-Out"
          variant="primary"
        />
        <button
          data-testid="coachBtn"
          icon="DefaultAvatarBold"
          style="width: 44px; height: 44px;"
          title="Disable Coach Mode"
          variant="primary"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;